### ISSUE TITLE
Expose CLI invocation reset

### ISSUE THREAD

**Issue by LecrisUT** [teemtee/tmt#2903]
I've found that in order to test tmt click command, we need to reset the cli invocation as:
https://github.com/teemtee/tmt/blob/78742f93789da5bbc13db3713cd6de3e8f744443/tests/__init__.py#L8-L29

For external plugin support it would be good to expose this function in `tmt.utils` or have a generic `tmt.testing` to simplify writing tests. I can try to gather some more ideas for the latter as I stumble upon them if that is the preferred approach. So far I have:
- test fixture for `cli_runner`, maybe as `tmt_cli`
- test fixture to initialize or read a tmt tree. How to use this might be quite tricky since most of the api and testing goes through the cli wrappers, but maybe it can be useful in the future for preparing various other steps and only testing one step :thinking: 
- lint helper
- `caplog` helper (maybe can already use the base one?)
- mock `tmt.utils.Command.run`. Probably that's the only mock helper needed?

**Comment by LecrisUT** [comment]
Actually a mock for `Command.run` would really be useful. I've been trying with `flexmock`, `unittest.mock` and `pytest-subprocess. So far unsuccessful

**Comment by happz** [comment]
> I've found that in order to test tmt click command, we need to reset the cli invocation as:

Correct. Unfortunate effect of how CLI options are stored. I for one am willing to live with `reset_common()` rather than trying to remove `cli_invocation` fields, because it was very painful to get it working *and* preserve the command-line behavior expected by existing tmt users, and I would happily focus on other parts of tmt, not touching this ever again, till the winter of 2047 when 1. I would be retired, and 2. every other feature and bugfix would be completed...

> For external plugin support it would be good to expose this function in `tmt.utils` or have a generic `tmt.testing` to simplify writing tests.

Makes sense. tmt does not ship with "testing tmt" library of helpers, it would make sense to offer one, there are several helpers and fixtures already available.

> I can try to gather some more ideas for the latter as I stumble upon them if that is the preferred approach. So far I have:
> 
> * test fixture for `cli_runner`, maybe as `tmt_cli`

+1, `CLIRunner` wrapped as a fixture would be nice.

> * test fixture to initialize or read a tmt tree. How to use this might be quite tricky since most of the api and testing goes through the cli wrappers, but maybe it can be useful in the future for preparing various other steps and only testing one step 🤔

For unit tests it would be indeed helpful.

> * lint helper

No idea what this would do.

> * `caplog` helper (maybe can already use the base one?)

tmt tests can use `assert_log` & co., bundling them into a `caplog` wrapper (`caplog.assert_log(message='foo'))`) would also work, but that can be an improvement once `assert_log` is available.

> * mock `tmt.utils.Command.run`. Probably that's the only mock helper needed?

> Actually a mock for `Command.run` would really be useful. I've been trying with `flexmock`, `unittest.mock` and [`pytest-subprocess`](https://github.com/aklajnert/pytest-subprocess/issues/155). So far unsuccessful


Can you elaborate on what are you looking for here, what kind of features? Because when I read "a mock for `Command.run`", I imagine simple `MagicMock` or similar - you obviously need something different than the very basic monkeypatch of `Command.run`.

**Comment by psss** [comment]
Looks like there are some unanswered questions, @LecrisUT?

**Comment by LecrisUT** [comment]
> > For external plugin support it would be good to expose this function in `tmt.utils` or have a generic `tmt.testing` to simplify writing tests.
> 
> Makes sense. tmt does not ship with "testing tmt" library of helpers, it would make sense to offer one, there are several helpers and fixtures already available.

Still relevant, and an easy step forward

> +1, `CLIRunner` wrapped as a fixture would be nice.

And we have this now :)

> Because when I read "a mock for `Command.run`", I imagine simple `MagicMock` or similar - you obviously need something different than the very basic monkeypatch of `Command.run`.

No, that's about it, just making it more automatic. `flexmock` has stuff like making sure a particular command was executed and how many times it was. and `pytest-subprocess` is neat in that you do not need to navigate to the specific instance to mock things, and a similar design would be nice here.

**Comment by thrix** [comment]
From triage: the main thing is the `testing tmt" library of helpers`
